### PR TITLE
Feat: Added localization support to create-hint package.

### PR DESCRIPTION
**Pull request by jaspreet57**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)
Added localization support to `create-hint` package.
Now if new hint is create in main repo (is official), it will have all necessary `i18n` scripts and files.

* `meta` will have `getDescription` and `getName` method
* `_locale` folder will get created with hint `description` and `name` for `en` language.

### CODE REVIEW

File: packages/hint/docs/contributor-guide/guides/create-custom-hint.md [hunk 1]
```diff
@@ -24,9 +24,15 @@ a resource file.
...
-slighthly different (references and task will change a bit).
+>**Note for core hints**: If you are creating a hint for the main repo you should
+>follow the below steps:
+>
+>* Make sure to run `yarn` and `yarn build` in root directory.
+>* Then go to `create-hint` package ( `cd packages/create-hint` ).
+>* Make executable release. run `yarn webpack`.
+>* Now run `npm create hint` from `packages` folder.
+>
+>The generated files will be slighthly different (references and task will change a bit).
```
Comment (antross) [inline packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: The length of this line is causing the PR to fail during markdown linting in CI. Just need to wrap it to reduce the length below 80 characters.

Comment (jaspreet57) [reply packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: Oh ok. Thanks. I am fixing it.

File: packages/create-hint/src/templates/meta.ts.hbs [hunk 2]
```diff
@@ -1,21 +1,43 @@
...
+import { getMessage } from './i18n.import';{{/if}}
 
 const meta: HintMetadata = {
     docs: {
-        category: Category.{{category}},
-        description: `{{description}}`
+        category: Category.{{hint.category}},{{#if packageData.official}}
+        description: getMessage('{{#if packageData.isMulti}}{{hint.normalizedName}}_{{/if}}description', 'en'),
+        name: getMessage('{{#if packageData.isMulti}}{{hint.normalizedName}}_{{/if}}name', 'en'){{else}}
+        description: `{{hint.description}}`,
```
Comment (sarvaje) [inline packages/create-hint/src/templates/meta.ts.hbs]: any reason to use ` instead of '?

Comment (jaspreet57) [reply packages/create-hint/src/templates/meta.ts.hbs]: Not any specific reason. We can remove this one.

Comment (jaspreet57) [reply packages/create-hint/src/templates/meta.ts.hbs]: It was already there earlier, So I just left it as it is. But I think, it should not have any role here.

Comment (sarvaje) [reply packages/create-hint/src/templates/meta.ts.hbs]: I was just curious. The only thing is that we should escape the character used (I don't know if we are already doing that or not)

File: packages/create-hint/src/templates/locales-messages.json.hbs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+    {{#each hints}}
+    "{{#if ../isMulti}}{{normalizedName}}_{{/if}}description": {
```
Comment (sarvaje) [inline packages/create-hint/src/templates/locales-messages.json.hbs]: We are not using the `normalizedName` here. We are using the `camelcase`, I think we already have a helper for that "toCamelCase"

File: packages/create-hint/src/templates/meta.ts.hbs [hunk 4]
```diff
@@ -1,21 +1,43 @@
...
+{{#if packageData.official}}
+
+import { getMessage } from './i18n.import';{{/if}}
 
 const meta: HintMetadata = {
     docs: {
-        category: Category.{{category}},
-        description: `{{description}}`
+        category: Category.{{hint.category}},{{#if packageData.official}}
+        description: getMessage('{{#if packageData.isMulti}}{{hint.normalizedName}}_{{/if}}description', 'en'),
```
Comment (sarvaje) [inline packages/create-hint/src/templates/meta.ts.hbs]: Same as before, it is not `normalizedName`. There are more places where you need to change it.

File: packages/hint/docs/contributor-guide/guides/create-custom-hint.md (hunk 1 above)
Comment (jaspreet57) [reply packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: done

File: packages/create-hint/src/templates/meta.ts.hbs (hunk 2 above)
Comment (jaspreet57) [reply packages/create-hint/src/templates/meta.ts.hbs]: Hmm right. Checking now if we are escaping or not. However backtick should be still safe to use in such cases.

File: packages/create-hint/src/templates/locales-messages.json.hbs (hunk 3 above)
Comment (sarvaje) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: You can see an example here: https://github.com/webhintio/hint/blob/master/packages/hint-typescript-config/src/_locales/en/messages.json

File: packages/create-hint/src/templates/meta.ts.hbs (hunk 2 above)
Comment (jaspreet57) [reply packages/create-hint/src/templates/meta.ts.hbs]: we are doing `escape` while creating `NewHint`
```
this.description = escapeSafeString(hintData.description);
```

Should I keep single-quote?

Comment (sarvaje) [reply packages/create-hint/src/templates/meta.ts.hbs]: nah, it is ok.

File: packages/create-hint/src/templates/locales-messages.json.hbs (hunk 3 above)
Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Yes right. good catch. Thanks. I didn't see for that long names.

So should I use `toCamelCase` helper on `normalizedName` or any other name?

Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: I got it. It is helper registered there...

File: packages/create-hint/src/templates/locales-messages.json.hbs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+    {{#each hints}}
+    "{{#if ../isMulti}}{{toCamelCase normalizedName}}_{{/if}}description": {
+        "description": "Metadata description",
+        "message": "{{description}}"
+    },
+    "{{#if ../isMulti}}{{toCamelCase normalizedName}}_{{/if}}name": {
+        "description": "Metadata name",
+        "message": "{{normalizedName}}"
```
Comment (jaspreet57) [inline packages/create-hint/src/templates/locales-messages.json.hbs]: @sarvaje key name is in `camelCase`. But actual name value seems to be `space` delimited with first char capital. Right?
Do we have helper for that in utils? For me it seems like `normalize-string-with-delimiter` will work... But should I create one for same?

Comment (sarvaje) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: can you give me an example?

Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Like in `hint-axe` locale file...
https://github.com/webhintio/hint/blob/master/packages/hint-axe/src/_locales/en/messages.json

see `nameRoleValue_name` > `message` value
```
"nameRoleValue_name": {
        "description": "Metadata name",
        "message": "Name role value"
 },
```

I just see similar pattern in many hints though it is just a value, user can configure it.
I was asking for default value when hint gets created.

Comment (sarvaje) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Ahhh, yes, the actual name is like you said. I was talking only about the key. Sorry for the misunderstanding

Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: ok that is fine. So my changes are done. Please have a review once.

File: packages/hint/docs/contributor-guide/guides/create-custom-hint.md [hunk 6]
```diff
@@ -32,7 +32,8 @@ the CLI.
 >* Make executable release. run `yarn webpack`.
 >* Now run `npm create hint` from `packages` folder.
 >
->The generated files will be slighthly different (references and task will change a bit).
+>The generated files will be
+>slightly different (references and task will change a bit).
```
Comment (antross) [inline packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: If you can, you may want to split this markdown fix into its own commit. 

Context: When we're ready to merge this I'd like to squash down to two commits - one for everything under `packages/create-hint` with the `New:` label and another for the doc updates under `packages/hint` with the `Docs:` label. This will help our release script bump versions appropriately.

Comment (jaspreet57) [reply packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: Ahh yes right. Let me do this.

**Comment by antross** [comment]
@sarvaje Do you recall the intent behind the `templates` versus `shared-templates` folders? Perhaps I'm missing something, but I don't see a logical separation between how the two are being used.

File: packages/hint/docs/contributor-guide/guides/create-custom-hint.md (hunk 6 above)
Comment (jaspreet57) [reply packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: @antross I have re-based this PR to 2 commits. One is for docs and another one for updates.

**Comment by jaspreet57** [comment]
> Current changes look good 
> 
> One new request: I noticed we don't currently emit a `context.report` example in the generated `hint.ts`. I think that'd be good to add as part of this PR so we can show how it should be localized when generating an official hint. We should also include the string for that message in the generated `_locales/en/messages.json` file in that case.

Good Idea :)... I will add that..

**Comment by molant** [comment]
Before breaking everything into packages the create- tasks shared some templates.
To speed up the process we decided to copy paste the folder instead of finding something else.

That's the history behind. It should be safe to merge both folders even though there will still be some duplication among the create packages.

Maybe a "good first issue"?

________________________________

File: packages/hint/docs/contributor-guide/guides/create-custom-hint.md (hunk 6 above)
Comment (antross) [reply packages/hint/docs/contributor-guide/guides/create-custom-hint.md]: Thanks! One minor tweak, let's use `New:` instead of `Chore:` for the commit with the functionality updates. We're adding support for localization of official hints to `create-hint` so I'd consider it a new feature.

**Comment by jaspreet57** [comment]
> > Current changes look good
> > One new request: I noticed we don't currently emit a `context.report` example in the generated `hint.ts`. I think that'd be good to add as part of this PR so we can show how it should be localized when generating an official hint. We should also include the string for that message in the generated `_locales/en/messages.json` file in that case.
> 
> Good Idea :)... I will add that..

@antross Added i18n support in `hint.ts` code also for official packages. `context.report` is now using `getMessage` of `./i18n-imports`.
I just had to pass `{ hint, packageData }` to `hint.ts.hbs` template to differentiate between hint and root data.

And `New` tag also added to feat commits :)

File: packages/create-hint/src/templates/locales-messages.json.hbs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    },
+    "{{#if ../isMulti}}{{toCamelCase normalizedName}}_{{/if}}reportMessage": {
```
Comment (antross) [inline packages/create-hint/src/templates/locales-messages.json.hbs]: Other than `name` and `description`, we don't generally prefix messages in `messages.json`, even for multi-hints. The name and description metadata is an exception because we need those names to follow a well-defined format so we can use them when generating documentation.
```suggestion
    "reportMessage": {
```

File: packages/create-hint/src/templates/partial-event-code.hbs [hunk 8]
```diff
@@ -12,5 +12,9 @@ debug(`Validating hint {{name}}`);
  */
 
 if (Math.ceil(Math.random()) === 0) {
+    {{#if packageData.official}}
+    context.report(resource, getMessage('{{#if packageData.isMulti}}{{toCamelCase hint.normalizedName}}_{{/if}}reportMessage', context.language));
```
Comment (antross) [inline packages/create-hint/src/templates/partial-event-code.hbs]:
```suggestion
    context.report(resource, getMessage('reportMessage', context.language));
```

File: packages/create-hint/src/templates/locales-messages.json.hbs (hunk 7 above)
Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: @antross  that is right if user create new messages in locales file himself. But here I am auto-generating, and for demo purpose, I am keeping name of key as `reportMessage`. Now what should I do for multiple hints here? Key name has to be prefixed with something.

Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Or should I just add only one message as `reportMessage` and use it in all hints of multi hint package

Comment (antross) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Good point. I like your suggestion of using a common message in this case. Otherwise I'm concerned new users would think they have to prefix all messages.

Comment (jaspreet57) [reply packages/create-hint/src/templates/locales-messages.json.hbs]: Thanks @antross. And I have made the change. Please review last commit.

**Comment by jaspreet57** [comment]
> Looks good.
> 
> Can you squash this back down to two commits? Then I'll go ahead and merge. Thanks!

Rebased and squashed. And had fun :)
